File: sketch/src/main/java/com/github/panpf/sketch/util/pool/AttributeStrategy.java
package com.github.panpf.sketch.util.pool;

import android.graphics.Bitmap;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.github.panpf.sketch.util.BitmapUtilsKt;

import java.util.Locale;

/**
 * A strategy for reusing bitmaps that requires any returned bitmap's dimensions to exactly match those request.
 */
public class AttributeStrategy implements LruPoolStrategy {

    @NonNull
    private final KeyPool keyPool = new KeyPool();
    @NonNull
    private final GroupedLinkedMap<Key, Bitmap> groupedMap = new GroupedLinkedMap<>();

    @Override
    public void put(@NonNull Bitmap bitmap) {
        final Key key = keyPool.get(bitmap.getWidth(), bitmap.getHeight(), bitmap.getConfig());
        groupedMap.put(key, bitmap);
    }

    @Nullable
    @Override
    public Bitmap get(int width, int height, @Nullable Bitmap.Config config) {
        final Key key = keyPool.get(width, height, config);
        return groupedMap.get(key);
    }

    @Override
    public boolean exist(int width, int height, @Nullable Bitmap.Config config) {
        final Key key = keyPool.get(width, height, config);
        return groupedMap.exist(key);
    }

    @Override
    public boolean exist(Bitmap bitmap) {
        final Key key = keyPool.get(bitmap.getWidth(), bitmap.getHeight(), bitmap.getConfig());
        return groupedMap.exist(key, bitmap);
    }

    @Nullable
    @Override
    public Bitmap removeLast() {
        return groupedMap.removeLast();
    }

    @NonNull
    @Override
    public String logBitmap(@NonNull Bitmap bitmap) {
        return getBitmapString(bitmap);
    }

    @NonNull
    @Override
    public String logBitmap(int width, int height, @Nullable Bitmap.Config config) {
        return getBitmapString(width, height, config);
    }

    @Override
    public int getSize(@NonNull Bitmap bitmap) {
        return BitmapUtilsKt.getAllocationByteCountCompat(bitmap);
    }

    @NonNull
    @Override
    public String toString() {
        return "AttributeStrategy(" + groupedMap + ")";
    }

    @NonNull
    private static String getBitmapString(@NonNull Bitmap bitmap) {
        return getBitmapString(bitmap.getWidth(), bitmap.getHeight(), bitmap.getConfig());
    }

    @NonNull
    private static String getBitmapString(int width, int height, @Nullable Bitmap.Config config) {
        return String.format(Locale.getDefault(), "[%dx%d](%s)", width, height, config);
    }

    private static class KeyPool extends BaseKeyPool<Key> {

        @NonNull
        public Key get(int width, int height, @Nullable Bitmap.Config config) {
            Key result = get();
            result.init(width, height, config);
            return result;
        }

        @NonNull
        @Override
        protected Key create() {
            return new Key(this);
        }
    }

    private static class Key implements Poolable {

        @NonNull
        private final KeyPool pool;
        private int width;
        private int height;
        @Nullable
        private Bitmap.Config config;

        public Key(@NonNull KeyPool pool) {
            this.pool = pool;
        }

        public void init(int width, int height, @Nullable Bitmap.Config config) {
            this.width = width;
            this.height = height;
            this.config = config;
        }

        @Override
        public boolean equals(Object o) {
            if (o instanceof Key) {
                Key other = (Key) o;
                return width == other.width && height == other.height && config == other.config;
            }
            return false;
        }

        @Override
        public int hashCode() {
            int result = width;
            result = 31 * result + height;
            result = 31 * result + (config != null ? config.hashCode() : 0);
            return result;
        }

        @NonNull
        @Override
        public String toString() {
            return getBitmapString(width, height, config);
        }

        @Override
        public void offer() {
            pool.offer(this);
        }
    }
}

File: sketch-core/src/main/java/com/github/panpf/sketch/util/pool/AttributeStrategy.kt
package com.github.panpf.sketch.util.pool

import android.graphics.Bitmap
import android.graphics.Bitmap.Config
import com.github.panpf.sketch.util.allocationByteCountCompat
import java.util.Locale

/**
 * A strategy for reusing bitmaps that requires any returned bitmap's dimensions to exactly match those request.
 */
class AttributeStrategy : LruPoolStrategy {

    private val keyPool = KeyPool()
    private val groupedMap = GroupedLinkedMap<Key, Bitmap>()

    override fun put(bitmap: Bitmap) {
        val key = keyPool[bitmap.width, bitmap.height, bitmap.config]
        groupedMap.put(key, bitmap)
    }

    override fun get(width: Int, height: Int, config: Config): Bitmap? {
        val key = keyPool[width, height, config]
        return groupedMap[key]
    }

    override fun exist(width: Int, height: Int, config: Config): Boolean {
        val key = keyPool[width, height, config]
        return groupedMap.exist(key)
    }

    override fun removeLast(): Bitmap? {
        return groupedMap.removeLast()
    }

    override fun logBitmap(bitmap: Bitmap): String {
        return getBitmapString(bitmap)
    }

    override fun logBitmap(width: Int, height: Int, config: Config): String {
        return getBitmapString(width, height, config)
    }

    override fun getSize(bitmap: Bitmap): Int {
        return bitmap.allocationByteCountCompat
    }

    override fun toString(): String {
        return "AttributeStrategy($groupedMap)"
    }

    private class KeyPool : BaseKeyPool<Key?>() {
        operator fun get(width: Int, height: Int, config: Config?): Key {
            val result = get()
            result!!.init(width, height, config)
            return result
        }

        override fun create(): Key {
            return Key(this)
        }
    }

    private class Key(private val pool: KeyPool) : Poolable {
        private var width = 0
        private var height = 0
        private var config: Config? = null
        fun init(width: Int, height: Int, config: Config?) {
            this.width = width
            this.height = height
            this.config = config
        }

        override fun equals(other: Any?): Boolean {
            if (other is Key) {
                return width == other.width && height == other.height && config == other.config
            }
            return false
        }

        override fun hashCode(): Int {
            var result = width
            result = 31 * result + height
            result = 31 * result + if (config != null) config.hashCode() else 0
            return result
        }

        override fun toString(): String {
            return getBitmapString(width, height, config)
        }

        override fun offer() {
            pool.offer(this)
        }
    }

    companion object {
        private fun getBitmapString(bitmap: Bitmap): String {
            return getBitmapString(bitmap.width, bitmap.height, bitmap.config)
        }

        private fun getBitmapString(width: Int, height: Int, config: Config?): String {
            return String.format(Locale.getDefault(), "[%dx%d](%s)", width, height, config)
        }
    }
}